### PR TITLE
Remove Vulkan timeline semaphore emulation code.

### PR DESCRIPTION
**Pull request by ScottTodd**
Part of https://github.com/google/iree/issues/4494 for migrating to Vulkan 1.2
Related to https://github.com/google/iree/issues/2907 for enabling tests that were previously failing with emulation

This requires that all of our test configurations support timeline semaphores:
- [x] Desktop SwiftShader CPU (may require a version bump)
  * [`bazel_linux_x86-swiftshader_core` failure logs](https://source.cloud.google.com/results/invocations/08badaac-d47e-42e8-8efa-7f7f2074d0a3/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log): `/install-swiftshader/swiftshader/src/Vulkan/VkSemaphore.cpp:123 WARNING: nextInfo->sType = 1000207002` -- blocked on https://github.com/google/iree/issues/7419
  * [`cmake_linux_x86-swiftshader` mostly passing logs](https://source.cloud.google.com/results/invocations/7b9ab706-94f6-4e6d-a6c9-2751df6477e0/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log): just semaphore_test timeouts in a test I tried to enable again
- [x] Desktop NVIDIA GPU (if we're testing there; may require a driver update)
  * [`cmake_linux_x86-turing` logs](https://source.cloud.google.com/results/invocations/b8c4f44a-ca14-43ee-860a-14c50033629d/targets): success
- [x] Android GPU (may require more recent hardware up to date on OS patches)
  * [`iree-android-arm64-v8a` logs](https://buildkite.com/iree/iree-android-arm64-v8a/builds/5526) - Galaxy S10 passes tests, Galaxy S20 timed out (?), ~~Pixel 4 fails them (can we switch to a newer Pixel phone for testing?)~~
    * We no longer test correctness on Pixel 4?
  * ~~[`iree-benchmark` logs](https://buildkite.com/iree/iree-benchmark/builds/1779) - Pixel 4 failed, Pixel 6 succeeded, no summary comment for easy analysis~~
    * Fixed by disabling Vulkan benchmarks on Pixel 4 (we still benchmark on Pixel 6 and Moto Edge X30)

### CODE REVIEW

**Comment by antiagainst** [comment]
Pixel 4 will be problematic; given we have an outdated driver there. Recent Qualcomm phones does [start to contain](https://vulkan.gpuinfo.org/listdevicescoverage.php?platform=android&extension=VK_KHR_timeline_semaphore) newer drivers that support timeline semaphores. I bought [one](https://vulkan.gpuinfo.org/displayreport.php?id=13441) and will evaluate how stable it is for benchmarking. It's still under shipping. Likely will take some extra weeks before we can finalize the choice of a phone and buy them for the lab. 

Trying to understand the timeline here. How soon are we expected to land this?

**Comment by ScottTodd** [comment]
> Trying to understand the timeline here. How soon are we expected to land this?

Pessimistically: within a year? It's already been one year since we filed https://github.com/google/iree/issues/4494 and started looking at this :P

I put this PR together to see what would break, and it seems like we're actually pretty close to being able to make the flip. @benvanik is doing more real work with timeline semaphores now and being able to move the HAL the rest of the way to pure C with native timeline semaphore support assumed would definitely simplify things.

**Comment by benvanik** [comment]
It's hampering work on cleaning up semaphore and resource lifetime handling in the HAL - I've been putting off doing that for a year because of it and have run out of time that I can put it off. In the current form it may need a non-trivial rewrite that I'd much rather lose device support for devices we don't and won't ever ship on than keep kicking the can on getting things right :)

**Comment by benvanik** [comment]
(we may be able to switch to the newer helpers I'm adding and maintain emulation for coverage but it wouldn't be with this code - so it's more of an issue of how to properly rewrite all of semaphore handling out from under the HAL while also maintaining this in a functioning state - that's been the mental block for me)

**Comment by benvanik** [comment]
Sorry - that was unclear - there's really two things here:
If we don't need to keep the existing emulation active during the reworking because we can use pixel 6 instead of pixel 4 that speeds things up/helps my mental block in getting the compiler/HAL interfaces/other backends in a good state.
If we never need non-timeline-semaphore paths again such that we don't need to hook emulation back up for Vulkan that reduces the amount of work required. Since we aren't shipping Android/Vulkan any time soon we can defer that work.

So we can still need emulation but if we don't need it _now_ then that's a good reason to yank it.

**Comment by ScottTodd** [comment]
Benchmarks completed on the Pixel 6 but failed (as expected I guess :P) on the Pixel 4: https://buildkite.com/iree/iree-benchmark/builds/1779. Looks like no comment gets posted with summary information if one phone fails.

Worth commenting out the Pixel 4 benchmarks and running again to easily see if this affects performance? https://github.com/google/iree/blob/0e39d8edbff08f47c2dc3c0f6d26bc27d0313e8d/build_tools/buildkite/cmake/android/arm64-v8a/benchmark2.yml#L26

**Comment by benvanik** [comment]
If it doesn't run on the pixel 6 then we have our answer as to whether we can get away without it :P

**Comment by ScottTodd** [comment]
Err, I mean the Pixel 6 completed all benchmark runs successfully, but the Pixel 4 failed. Since one failed, there was no comment posted with summary data.

**Comment by antiagainst** [comment]
@ScottTodd: Certainly no need to wait another year! :D As said there are new devices having native support; just need to find the right one for benchmarking. (Galaxy S20 really bit us hard; don't want to have another episode of that. So I'm cautious.) 

@benvanik: What you said makes sense to me! I think we will need the emulation for supporting the broad range of phones given the notorious Android fragmentation issue. But that's quite far so makes sense to not anchor on it too much. The value of having it right now is being able to get benchmark numbers on Adreno GPUs. Though, I don't want to waste more of your time to handholding it given it's really close that we can find a phone with native support. I'm more from the perspective of understanding how soon this will happen and whether I can help to speed up by prioritizing certain things. (It depends on the logistics to order phones and get them shipped and set up, which is a bit uncontrollable by me. But I guess Google have the power of $$ to speed up a bit. Or at least I can order more than one and check them in parallel maybe. ;-P)

Let's sync up via GVC tomorrow so we can align up. :)

**Comment by benvanik** [comment]
Agreed! I'm going to put some thought into a way to swap it out with the new stuff - this was more of a speculative "if we could avoid having that PR being 10k LOC (and instead be like 5k) that'd be nice" :P
My hope is that the newer stuff will be easier to implement and be useful for implementations where we have to emulate (CUDA/ROCM). But I've been promising this working for so long no one should trust me until it lands :P

File: build_tools/buildkite/cmake/android/arm64-v8a/benchmark2.yml [hunk 1]
```diff
@@ -34,7 +34,7 @@ steps:
       - "tar -xzvf benchmark-suites-${BUILDKITE_BUILD_NUMBER}.tgz"
       - "tar -xzvf iree-android-tools-${BUILDKITE_BUILD_NUMBER}.tgz"
       - "tar -xzvf tracy-capture-80f6a93d.tgz"
-      - "python3 build_tools/benchmarks/run_benchmarks_on_android.py --pin-cpu-freq --pin-gpu-freq --normal_benchmark_tool_dir=build-android/tools/ --traced_benchmark_tool_dir=build-android-trace/tools/ --trace_capture_tool=tracy-capture -o benchmark-results-pixel-4-${BUILDKITE_BUILD_NUMBER}.json --capture_tarball=trace-captures-pixel-4-${BUILDKITE_BUILD_NUMBER}.tgz --verbose build-host/"
+      - "python3 build_tools/benchmarks/run_benchmarks_on_android.py --pin-cpu-freq --pin-gpu-freq --normal_benchmark_tool_dir=build-android/tools/ --traced_benchmark_tool_dir=build-android-trace/tools/ --trace_capture_tool=tracy-capture -o benchmark-results-pixel-4-${BUILDKITE_BUILD_NUMBER}.json --capture_tarball=trace-captures-pixel-4-${BUILDKITE_BUILD_NUMBER}.tgz --driver_filter_regex='(?!vulkan).*' --verbose build-host/"
```
Comment (ScottTodd) [inline build_tools/buildkite/cmake/android/arm64-v8a/benchmark2.yml:37]: Note that this disables Vulkan benchmarks on Pixel 4

**Review by ScottTodd — COMMENTED** [review]
Synced to head. This passes presubmits.

**Review by antiagainst — APPROVED** [review]

**Pull request merged**